op-monitorism/transaction_monitor/monitor.go

```diff
@@ -0,0 +1,161 @@
+package transaction_monitor
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/prometheus/client_golang/prometheus"
+    "github.com/ethereum-optimism/optimism/op-service/metrics"
+)
+
+const (
+	MetricsNamespace = "tx_mon"
+)
+
+type Monitor struct {
+	log           log.Logger
+	client        *ethclient.Client
+	watchConfigs  map[common.Address]WatchConfig // map of watch address to its config
+
+	// metrics
+	transactions         *prometheus.CounterVec
+	unauthorizedTx      *prometheus.CounterVec
+	thresholdExceededTx *prometheus.CounterVec
+	unexpectedRpcErrors *prometheus.CounterVec
+}
+
+
+func NewMonitor(ctx context.Context, log log.Logger, m metrics.Factory, cfg CLIConfig) (*Monitor, error) {
+	log.Info("creating transaction monitor")
+	client, err := ethclient.Dial(cfg.L1NodeUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	watchConfigs := make(map[common.Address]WatchConfig)
+	for _, config := range cfg.WatchConfigs {
+		watchConfigs[config.Address] = config
+		log.Info("monitoring address", 
+			"address", config.Address, 
+			"allowlist_size", len(config.AllowList))
+	}
+
+	return &Monitor{
+		log:          log,
+		client:       client,
+		watchConfigs: watchConfigs,
+
+		transactions: m.NewCounterVec(prometheus.CounterOpts{
+			Namespace: MetricsNamespace,
+			Name:      "transactions_total",
+			Help:      "total number of transactions",
+		}, []string{"watch_address", "from", "to", "status"}),
+
+		unauthorizedTx: m.NewCounterVec(prometheus.CounterOpts{
+			Namespace: MetricsNamespace,
+			Name:      "unauthorized_transactions_total",
+			Help:      "number of transactions from unauthorized addresses",
+		}, []string{"watch_address", "from"}),
+
+		thresholdExceededTx: m.NewCounterVec(prometheus.CounterOpts{
+			Namespace: MetricsNamespace,
+			Name:      "threshold_exceeded_transactions_total",
+			Help:      "number of transactions exceeding allowed threshold",
+		}, []string{"watch_address", "from", "threshold"}),
+
+		unexpectedRpcErrors: m.NewCounterVec(prometheus.CounterOpts{
+			Namespace: MetricsNamespace,
+			Name:      "unexpected_rpc_errors_total",
+			Help:      "number of unexpected rpc errors",
+		}, []string{"section", "name"}),
+	}, nil
+}
+
+func (m *Monitor) Run(ctx context.Context) {
+	blockNumber, err := m.client.BlockNumber(ctx)
```

Comment:
Should keep track of last block we've scanned as well as latest block and attempt to scan all blocks in between